mem.go

```diff
@@ -571,40 +573,51 @@ func (e *Encoder) Encode(m *Message) error {
 	if nsegs == 0 {
 		return errMessageEmpty
 	}
-	sizes, err := m.segmentSizes()
-	if err != nil {
-		return err
-	}
-	maxSeg := uint32(nsegs - 1)
-	hdrSize := streamHeaderSize(maxSeg)
-	if uint64(cap(e.hdrbuf)) < hdrSize {
-		e.hdrbuf = make([]byte, hdrSize)
+	if int64(cap(e.bufs)) < 1+nsegs {
+		e.bufs = make([][]byte, 1+nsegs)
 	} else {
-		e.hdrbuf = e.hdrbuf[:hdrSize]
+		e.bufs = e.bufs[:1+nsegs]
 	}
-	marshalStreamHeader(e.hdrbuf, sizes)
-	if err := e.write(e.hdrbuf); err != nil {
-		return err
+	if int64(cap(e.sizes)) < nsegs {
+		e.sizes = make([]Size, nsegs)
+	} else {
+		e.sizes = e.sizes[:nsegs]
 	}
 	for i := int64(0); i < nsegs; i++ {
 		s, err := m.Segment(SegmentID(i))
 		if err != nil {
 			return err
 		}
-		if err := e.write(s.data); err != nil {
-			return err
+		n := len(s.data)
```

Comment:
That would risk an overflow that the `int64(n) > int64(maxSize)` is designed to avoid. Size is a `uin32` but `len(x)` returns an `int` which can easily be larger than a `uint32` can handle.